File: 2021/day6.jl
using Test
using DataStructures

function part_one(input, days = 80)
    fish = parse.(Int, split(input, ","))
    c = counter(fish)

    for _day = 1:days
        zeros = c[0]
        for i = 0:7
            c[i] = c[i+1]
        end
        c[6] += zeros
        c[8] = zeros
    end

    return sum(c)
end

function part_two(input)
    return part_one(input, 256)
end

function test()
    example = """3,4,3,1,2"""

    @test part_one(example) == 5934
    @test part_two(example) == 26984457539
end

File: ruby/2023/day6.rb
require './commons/assert'

def dist(total, held)
  held * (total - held)
end

def part_one(input)
  times, distances = input.split("\n").map { |s| s.split(':')[1].split(' ').map(&:to_i) }
  races = times.zip(distances)
  races.map do |race|
    t, d = race
    (0..t).map { |held| dist(t, held) }.filter { |x| x > d }.count
  end.inject(:*)
end

def part_two(input)
  t, d = input.split("\n").map { |s| s.split(':')[1].gsub(' ', '').to_i }
  (0..t).map { |held| dist(t, held) }.filter { |x| x > d }.count
end

def test
  example = %(Time:      7  15   30
Distance:  9  40  200)

  assert part_one(example), 288
  assert part_two(example), 71_503
  0
end
